samples/net/sockets/big_http_download/prj_tls.conf

```diff
@@ -0,0 +1,48 @@
+# General config
+CONFIG_NEWLIB_LIBC=y
+CONFIG_MAIN_STACK_SIZE=4096
+CONFIG_NET_BUF_RX_COUNT=160
```

Comment:
Why such a big number? Seems to be pretty arbitrary and will raise the bar for RAM required beyond any reasonable figures. For example, in my testing, to connect to a handful of servers on the net, I had to set mbedTLS heap size to a figure (~60K IIRC) which led to total RAM usage of ~90K. With buf count like above, this sample won't be able to run even on boards with 128K.

Reply:
It's due to size of TLS chunks (16K). mbedTLS processes 16K chunk at once, and we use rx buffers to pass decrypted data to the application.

Reply:
1. Yes, the max size of TLS record is 16K. And indeed, that should be accumulated and processed at once by mbedTLS. That's why mbedTLS has *its own* heap, where *it* allocates input buffer for a TLS record. It's not related in to number of buffers in a networking stack mbedTLS integrates with - indeed, mbedTLS buffer can be fed piece by piece, with as small number of bytes as possible.

2. With the above info in mind, there's not even numerical relation between TLS record size 16K and 160 of Zephyr' net_buf's (which, with the default net_buf size, is 160*128=20K).

3. Another sample you converted, http_get, doesn't have this setting.

So, overall, this is a "magic" setting which nobody knows why is there, but causing issues for some targets (e.g. memory overflow) and which everyone is afraid to touch for the fear of breaking something. Multiple conf's we have in echo_server/echo_client is a tribute to such situation. (I tried to fix it up once, but gave up. But now I eyeball for attempts to introduce new such magic configs ;-). )

Reply:
There is no magic in here, let me reiterate:
>  we use rx buffers to pass **decrypted data** to the application

It has nothing to do with feeding mbedTLS with data. Once mbedTLS decrypts 16K chunk of data, decrypted data is packed into net_buf's and passed to the callback function. As this version has no separate thread for TLS processing, we have to pass all data at once, hence need for so many buffers. Just to clarify, I'm not 100% happy with it, and I guess we'll end up having separate thread for TLS if we'll stick to TLS at net_context level. Still I think it could be addresses later.

>  there's not even numerical relation between TLS record size 16K and 160 of Zephyr' net_buf's

128 buffers for decrypted data, 32 are left for ongoing communication.

> http_get, doesn't have this setting

http_get does not process so much data.

Reply:
> let me reiterate:
>    we use rx buffers to pass decrypted data to the application

I don't see a reason for doing that.

> Once mbedTLS decrypts 16K chunk of data, decrypted data is packed into net_buf's and passed to the callback function. As this version has no separate thread for TLS processing, we have to pass all data at once, hence need for so many buffers.

Again, I don't see a reason for doing that. Native Zephyr API passes data to apps in terms of packets, a packet is limited in size by MTU setting. Largest known realistic MTU is 1500 bytes, and an app may rely on that. But actually, a packet doesn't have to be of MTU size, a packet just have to have a single net_buf attached, and that's the minimum which can be passed via native Zephyr API. No 16K passing at once should be required. Nor any additional threads. And for socket API (which was selected as the official Zephyr networking API), any amount of data can be returned to an app.

Zstream API was a fully developed solution which required neither additional buffers (beyond already gigantic buffers which mbedTLS itself holds) nor threads. If it happens that your solution unavoidably requires additional major overheads (which 16K and/or extra thread are), that would be another illustration how layering violation bleeds thru everywhere and even precludes efficient technical implementation. (But again, in all fairness, I don't see why that would be the case, even for layer-violated implementation.)